src/transformers/models/detr/modeling_detr.py

```diff
@@ -0,0 +1,2239 @@
+                out.append((x, mask))
+            return out
+        
+        assert pixel_mask is not None
+        # get final feature map
+        feature_map = feature_maps[-1]
+        # we downsample the pixel_mask to match the shape of the feature map
+        mask = F.interpolate(pixel_mask[None].float(), size=feature_map.shape[-2:]).to(torch.bool)[0]
+        return feature_map, mask
+
+
+class Joiner(nn.Sequential):
+    def __init__(self, backbone, position_embedding):
+        super().__init__(backbone, position_embedding)
+
+    def forward(self, pixel_values, pixel_mask):
+        if self[0].return_intermediate_layers:
+            features = self[0](pixel_values, pixel_mask)
+            feature_map, pixel_mask = features[-1]
+            pos = self[1](feature_map, pixel_mask).to(feature_map.dtype)
+            
+            return features, pos
+
+        else:
+            # first, send pixel_values and pixel_mask through backbone to obtain updated feature_map and pixel_mask
+            feature_map, pixel_mask = self[0](pixel_values, pixel_mask)
+
+            # next, create position embeddings for the outputs of the outputs_dict
+            pos = self[1](feature_map, pixel_mask).to(feature_map.dtype)
+
+            return feature_map, pixel_mask, pos
+
+
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+    """
+    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
+    """
+    bsz, src_len = mask.size()
+    tgt_len = tgt_len if tgt_len is not None else src_len
+
+    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
+
+    inverted_mask = 1.0 - expanded_mask
+
+    return inverted_mask.masked_fill(inverted_mask.bool(), torch.finfo(dtype).min)
+
+
+class DetrSinePositionEmbedding(nn.Module):
+    """
+    This is a more standard version of the position embedding, very similar to the one used by the Attention is all you
+    need paper, generalized to work on images.
+    """
+
+    def __init__(self, embedding_dim=64, temperature=10000, normalize=False, scale=None):
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.temperature = temperature
+        self.normalize = normalize
+        if scale is not None and normalize is False:
+            raise ValueError("normalize should be True if scale is passed")
+        if scale is None:
+            scale = 2 * math.pi
+        self.scale = scale
+
+    def forward(self, pixel_values, pixel_mask):
+        x = pixel_values
+        mask = pixel_mask
+        assert mask is not None
+        y_embed = mask.cumsum(1, dtype=torch.float32)
+        x_embed = mask.cumsum(2, dtype=torch.float32)
+        if self.normalize:
+            eps = 1e-6
+            y_embed = y_embed / (y_embed[:, -1:, :] + eps) * self.scale
+            x_embed = x_embed / (x_embed[:, :, -1:] + eps) * self.scale
```

Comment:
```suggestion
            y_embed = y_embed / (y_embed[:, -1:, :] + 1e-6) * self.scale
            x_embed = x_embed / (x_embed[:, :, -1:] + 1e-6) * self.scale
```
If it's hard-coded, there is no point adding a line for it.

Reply:
Isn't it useful to know that this is an epsilon value?